File: Data Structures & Algorithms/course-schedule-ii/submission-0.java
class Solution {

    private boolean solve(int node, boolean[] vis, boolean[] currVis, Stack<Integer> st, List<List<Integer>> adj){

        if(currVis[node])
            return false;
        if(vis[node])
            return true;
        
        vis[node] = true;
        currVis[node] = true;

        for(Integer it: adj.get(node)){
            if (!solve(it, vis, currVis, st, adj))
                return false;
        }
        currVis[node] = false;
        st.push(node);
        return true;
    }

    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adj = new ArrayList<>();
        Stack<Integer> st = new Stack<>();
        int[] ans = new int[numCourses];
        boolean[] vis = new boolean[numCourses];
        boolean[] currVis = new boolean[numCourses];

        for(int i=0; i<numCourses; i++)
            adj.add(new ArrayList<>());

        for(int[] it: prerequisites){
            adj.get(it[1]).add(it[0]);
        }

        for(int i=0; i<numCourses; i++){
            if(!solve(i, vis, currVis, st, adj))
                return new int[]{};
        }
        
        int c=0;
        while(!st.empty()){
            int x = st.pop();
            ans[c] = x;
            c++;
        }

        return ans;
    }
}

File: Data Structures & Algorithms/cheapest-flight-path/submission-0.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        priority_queue<pair<int, pair<int, int>>, vector<pair<int, pair<int, int>>>,
                       greater<pair<int, pair<int, int>>>>
            pq;

        vector<vector<pair<int, int>>> adj(n);

        for(auto it: flights){
            int u = it[0];
            int v = it[1];
            int w = it[2];
            adj[u].push_back({v,w});
        }

        vector<bool> vis(n, false);

        pq.push({0, {src, 0}});
        vis[src] = true;

        while(!pq.empty()){
            auto node = pq.top();
            pq.pop();
        
            int nd = node.second.first;
            int cost = node.first;
            int stops = node.second.second;

            if(nd == dst)
                return cost;

            vis[nd] = true;

            for(auto it: adj[nd]){
                if(!vis[it.first] && stops <= k){
                    pq.push({cost + it.second, {it.first, stops+1}});
                }
            }
        }

        return -1;
    }
};
